In [1]:
import pandas as pd

In [2]:
df = pd.read_excel('test SQL.xlsx','Test 2')

In [3]:
df

WORK_FLOW_ID  APP_ID               STATUS                CHANGED_DATE  \
0      48299262     123                  101         2021-03-25 15:27:55   
1      48299263     159                20006         2021-03-01 10:00:15   
2      48299264     159                  103         2021-03-28 16:15:55   
3      48299265     123                 1031         2021-03-25 17:56:57   
4      48305540     173                 1031  2021-03-30 08:07:23.467000   
5      48305542     173                  102         2021-04-02 08:07:25   
6      48306435     168                 3015         2021-02-28 08:39:26   
7      48309321     159                 3010         2021-03-29 11:00:15   
8      48310905     168                  301         2021-03-04 10:45:05   
9      48313471     123                  302         2021-03-29 12:01:57   
10          NaN     NaN                  NaN                         NaN   
11          NaN     NaN                  NaN                         NaN   
12           ID  APP_ID         CREATED_DATE                 LOAN_AMOUNT   
13            1     123  2021-03-24 15:15:55                    40000000   
14            2     159  2021-02-24 09:15:55                    30000000   
15            3     173  2021-03-29 13:15:55                    55000000   
16            3     168  2021-02-24 11:15:55                    40000000   
17            5     761  2021-03-14 13:40:55                    30000000   
18            6     234  2021-03-02 15:40:55                    35000000   

                                           Unnamed: 4  \
0                                                 NaN   
1                                                 NaN   
2                                                 NaN   
3                                                 NaN   
4                                                 NaN   
5                                                 NaN   
6                                                 NaN   
7                                                 NaN   
8                                                 NaN   
9                                                 NaN   
10                                                NaN   
11                                                NaN   
12                                         UW_COMMENT   
13  @uw01@ Trả sale bổ sung chứng từ|@uw04@ UW yêu...   
14                          @uw04@ UW yêu cầu bổ sung   
15                                                NaN   
16                                                NaN   
17  @uw06@ Trả chỉnh sửa thông tin|@uw01@ Trả sale...   
18                                                NaN   

   1. Cho biết danh sách tổng số tiền cho vay được tạo trong tháng 3/2021  
0   2. Cho biết trạng thái mới nhất của mỗi hồ sơ ...                      
1      3. Cho biết tổng thời gian xử lý của hồ sơ 159                      
2   4. Cho biết code lý do trả về cuối cùng của cá...                      
3                                                 NaN                      
4                                                 NaN                      
5                                                 NaN                      
6                                                 NaN                      
7                                                 NaN                      
8                                                 NaN                      
9                                                 NaN                      
10                                                NaN                      
11                                                NaN                      
12                                                NaN                      
13                                                NaN                      
14                                                NaN                      
15                                                NaN                      
16                                               

In [4]:
df_change = df[:10]
df_create = df[13:19]

In [5]:
df_change

WORK_FLOW_ID APP_ID STATUS                CHANGED_DATE Unnamed: 4  \
0     48299262    123    101         2021-03-25 15:27:55        NaN   
1     48299263    159  20006         2021-03-01 10:00:15        NaN   
2     48299264    159    103         2021-03-28 16:15:55        NaN   
3     48299265    123   1031         2021-03-25 17:56:57        NaN   
4     48305540    173   1031  2021-03-30 08:07:23.467000        NaN   
5     48305542    173    102         2021-04-02 08:07:25        NaN   
6     48306435    168   3015         2021-02-28 08:39:26        NaN   
7     48309321    159   3010         2021-03-29 11:00:15        NaN   
8     48310905    168    301         2021-03-04 10:45:05        NaN   
9     48313471    123    302         2021-03-29 12:01:57        NaN   

  1. Cho biết danh sách tổng số tiền cho vay được tạo trong tháng 3/2021  
0  2. Cho biết trạng thái mới nhất của mỗi hồ sơ ...                      
1     3. Cho biết tổng thời gian xử lý của hồ sơ 159                      
2  4. Cho biết code lý do trả về cuối cùng của cá...                      
3                                                NaN                      
4                                                NaN                      
5                                                NaN                      
6                                                NaN                      
7                                                NaN                      
8                                                NaN                      
9                                                NaN

In [6]:
col_change = ['WORK_FLOW_ID','APP_ID','STATUS','CHANGED_DATE']

In [7]:
df_change = df_change[col_change]

In [8]:
df_change

WORK_FLOW_ID APP_ID STATUS                CHANGED_DATE
0     48299262    123    101         2021-03-25 15:27:55
1     48299263    159  20006         2021-03-01 10:00:15
2     48299264    159    103         2021-03-28 16:15:55
3     48299265    123   1031         2021-03-25 17:56:57
4     48305540    173   1031  2021-03-30 08:07:23.467000
5     48305542    173    102         2021-04-02 08:07:25
6     48306435    168   3015         2021-02-28 08:39:26
7     48309321    159   3010         2021-03-29 11:00:15
8     48310905    168    301         2021-03-04 10:45:05
9     48313471    123    302         2021-03-29 12:01:57

In [9]:
df_change['CHANGED_DATE'] = pd.to_datetime(df_change['CHANGED_DATE'])

<ipython-input-9-e232512a1586>:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_change['CHANGED_DATE'] = pd.to_datetime(df_change['CHANGED_DATE'])


In [10]:
df_change.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10 entries, 0 to 9
Data columns (total 4 columns):
 #   Column        Non-Null Count  Dtype         
---  ------        --------------  -----         
 0   WORK_FLOW_ID  10 non-null     object        
 1   APP_ID        10 non-null     object        
 2   STATUS        10 non-null     object        
 3   CHANGED_DATE  10 non-null     datetime64[ns]
dtypes: datetime64[ns](1), object(3)
memory usage: 448.0+ bytes


In [11]:
col_create_01 = ['WORK_FLOW_ID','APP_ID','STATUS','CHANGED_DATE','Unnamed: 4']
df_create = df_create[col_create_01]

In [12]:
col_create_02 = ['ID','APP_ID','CREATED_DATE','LOAN_AMOUNT','UW_COMMENT']

In [13]:
df_create.columns = col_create_02

In [14]:
df_create

ID APP_ID         CREATED_DATE LOAN_AMOUNT  \
13  1    123  2021-03-24 15:15:55    40000000   
14  2    159  2021-02-24 09:15:55    30000000   
15  3    173  2021-03-29 13:15:55    55000000   
16  3    168  2021-02-24 11:15:55    40000000   
17  5    761  2021-03-14 13:40:55    30000000   
18  6    234  2021-03-02 15:40:55    35000000   

                                           UW_COMMENT  
13  @uw01@ Trả sale bổ sung chứng từ|@uw04@ UW yêu...  
14                          @uw04@ UW yêu cầu bổ sung  
15                                                NaN  
16                                                NaN  
17  @uw06@ Trả chỉnh sửa thông tin|@uw01@ Trả sale...  
18                                                NaN

In [15]:
df_create.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 6 entries, 13 to 18
Data columns (total 5 columns):
 #   Column        Non-Null Count  Dtype 
---  ------        --------------  ----- 
 0   ID            6 non-null      object
 1   APP_ID        6 non-null      object
 2   CREATED_DATE  6 non-null      object
 3   LOAN_AMOUNT   6 non-null      object
 4   UW_COMMENT    3 non-null      object
dtypes: object(5)
memory usage: 372.0+ bytes


In [16]:
df_create['CREATED_DATE'] = pd.to_datetime(df_create['CREATED_DATE'])

In [17]:
df_create.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 6 entries, 13 to 18
Data columns (total 5 columns):
 #   Column        Non-Null Count  Dtype         
---  ------        --------------  -----         
 0   ID            6 non-null      object        
 1   APP_ID        6 non-null      object        
 2   CREATED_DATE  6 non-null      datetime64[ns]
 3   LOAN_AMOUNT   6 non-null      object        
 4   UW_COMMENT    3 non-null      object        
dtypes: datetime64[ns](1), object(4)
memory usage: 372.0+ bytes


In [18]:
df_change['month'] = df_change['CHANGED_DATE'].dt.month
df_create['month'] = df_create['CREATED_DATE'].dt.month

<ipython-input-18-77c70620fe1a>:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_change['month'] = df_change['CHANGED_DATE'].dt.month


In [19]:
df_create

ID APP_ID        CREATED_DATE LOAN_AMOUNT  \
13  1    123 2021-03-24 15:15:55    40000000   
14  2    159 2021-02-24 09:15:55    30000000   
15  3    173 2021-03-29 13:15:55    55000000   
16  3    168 2021-02-24 11:15:55    40000000   
17  5    761 2021-03-14 13:40:55    30000000   
18  6    234 2021-03-02 15:40:55    35000000   

                                           UW_COMMENT  month  
13  @uw01@ Trả sale bổ sung chứng từ|@uw04@ UW yêu...      3  
14                          @uw04@ UW yêu cầu bổ sung      2  
15                                                NaN      3  
16                                                NaN      2  
17  @uw06@ Trả chỉnh sửa thông tin|@uw01@ Trả sale...      3  
18                                                NaN      3

1. Cho biết danh sách tổng số tiền cho vay được tạo trong tháng 3/2021

In [20]:
df_sum = df_create.groupby('month')['LOAN_AMOUNT'].sum().reset_index()

In [21]:
df_sum

month  LOAN_AMOUNT
0      2     70000000
1      3    160000000

Vậy tổng số tiền cho vay được tạo trong tháng 3/2021 là 160000000

2. Cho biết trạng thái mới nhất của mỗi hồ sơ được tạo trong tháng 3


In [22]:
df_merge = pd.merge(df_create, df_change, on = ['APP_ID'], how = 'left')

In [23]:
df_merge

ID APP_ID        CREATED_DATE LOAN_AMOUNT  \
0   1    123 2021-03-24 15:15:55    40000000   
1   1    123 2021-03-24 15:15:55    40000000   
2   1    123 2021-03-24 15:15:55    40000000   
3   2    159 2021-02-24 09:15:55    30000000   
4   2    159 2021-02-24 09:15:55    30000000   
5   2    159 2021-02-24 09:15:55    30000000   
6   3    173 2021-03-29 13:15:55    55000000   
7   3    173 2021-03-29 13:15:55    55000000   
8   3    168 2021-02-24 11:15:55    40000000   
9   3    168 2021-02-24 11:15:55    40000000   
10  5    761 2021-03-14 13:40:55    30000000   
11  6    234 2021-03-02 15:40:55    35000000   

                                           UW_COMMENT  month_x WORK_FLOW_ID  \
0   @uw01@ Trả sale bổ sung chứng từ|@uw04@ UW yêu...        3     48299262   
1   @uw01@ Trả sale bổ sung chứng từ|@uw04@ UW yêu...        3     48299265   
2   @uw01@ Trả sale bổ sung chứng từ|@uw04@ UW yêu...        3     48313471   
3                           @uw04@ UW yêu cầu bổ sung        2     48299263   
4                           @uw04@ UW yêu cầu bổ sung        2     48299264   
5                           @uw04@ UW yêu cầu bổ sung        2     48309321   
6                                                 NaN        3     48305540   
7                                                 NaN        3     48305542   
8                                                 NaN        2     48306435   
9                                                 NaN        2     48310905   
10  @uw06@ Trả chỉnh sửa thông tin|@uw01@ Trả sale...        3          NaN   
11                                                NaN        3          NaN   

   STATUS            CHANGED_DATE  month_y  
0     101 2021-03-25 15:27:55.000      3.0  
1    1031 2021-03-25 17:56:57.000      3.0  
2     302 2021-03-29 12:01:57.000      3.0  
3   20006 2021-03-01 10:00:15.000      3.0  
4     103 2021-03-28 16:15:55.000      3.0  
5    3010 2021-03-29 11:00:15.000      3.0  
6    1031 2021-03-30 08:07:23.467      3.0  
7     102 2021-04-02 08:07:25.000      4.0  
8    3015 2021-02-28 08:39:26.000      2.0  
9     301 2021-03-04 10:45:05.000      3.0  
10    NaN                     NaT      NaN  
11    NaN                     NaT      NaN

In [24]:
df_create_month_3 = df_merge[df_merge.month_x == 3]

In [25]:
df_create_month_3

ID APP_ID        CREATED_DATE LOAN_AMOUNT  \
0   1    123 2021-03-24 15:15:55    40000000   
1   1    123 2021-03-24 15:15:55    40000000   
2   1    123 2021-03-24 15:15:55    40000000   
6   3    173 2021-03-29 13:15:55    55000000   
7   3    173 2021-03-29 13:15:55    55000000   
10  5    761 2021-03-14 13:40:55    30000000   
11  6    234 2021-03-02 15:40:55    35000000   

                                           UW_COMMENT  month_x WORK_FLOW_ID  \
0   @uw01@ Trả sale bổ sung chứng từ|@uw04@ UW yêu...        3     48299262   
1   @uw01@ Trả sale bổ sung chứng từ|@uw04@ UW yêu...        3     48299265   
2   @uw01@ Trả sale bổ sung chứng từ|@uw04@ UW yêu...        3     48313471   
6                                                 NaN        3     48305540   
7                                                 NaN        3     48305542   
10  @uw06@ Trả chỉnh sửa thông tin|@uw01@ Trả sale...        3          NaN   
11                                                NaN        3          NaN   

   STATUS            CHANGED_DATE  month_y  
0     101 2021-03-25 15:27:55.000      3.0  
1    1031 2021-03-25 17:56:57.000      3.0  
2     302 2021-03-29 12:01:57.000      3.0  
6    1031 2021-03-30 08:07:23.467      3.0  
7     102 2021-04-02 08:07:25.000      4.0  
10    NaN                     NaT      NaN  
11    NaN                     NaT      NaN

In [26]:
df_status_new = df_create_month_3.groupby('APP_ID')['CHANGED_DATE'].max().reset_index()

In [27]:
df_status_new

APP_ID        CHANGED_DATE
0     123 2021-03-29 12:01:57
1     173 2021-04-02 08:07:25
2     234                 NaT
3     761                 NaT

In [28]:
df_status = pd.merge(df_status_new, df_change, on = ['APP_ID', 'CHANGED_DATE'], how = 'left')

In [29]:
df_status

APP_ID        CHANGED_DATE WORK_FLOW_ID STATUS  month
0    123 2021-03-29 12:01:57     48313471    302    3.0
1    173 2021-04-02 08:07:25     48305542    102    4.0
2    234                 NaT          NaN    NaN    NaN
3    761                 NaT          NaN    NaN    NaN

Vậy trạng thái mới nhất của mỗi hồ sơ được tạo trong tháng 3 gồm:
    - Với hồ sơ có APP_ID là 123 và 173 thì trạng thái mới nhất lần lượt là 302 và 102.
    - Với hồ sơ có APP_ID là 234 và 761 thì không có sự thay đổi trạng thái
    
    

3. Cho biết tổng thời gian xử lý của hồ sơ 159


In [30]:
df_03 = df_merge[df_merge.APP_ID == 159]

In [31]:
df_03

ID APP_ID        CREATED_DATE LOAN_AMOUNT                 UW_COMMENT  \
3  2    159 2021-02-24 09:15:55    30000000  @uw04@ UW yêu cầu bổ sung   
4  2    159 2021-02-24 09:15:55    30000000  @uw04@ UW yêu cầu bổ sung   
5  2    159 2021-02-24 09:15:55    30000000  @uw04@ UW yêu cầu bổ sung   

   month_x WORK_FLOW_ID STATUS        CHANGED_DATE  month_y  
3        2     48299263  20006 2021-03-01 10:00:15      3.0  
4        2     48299264    103 2021-03-28 16:15:55      3.0  
5        2     48309321   3010 2021-03-29 11:00:15      3.0

In [32]:
total_process_159 = df_03['CHANGED_DATE'].max() - df_03['CREATED_DATE'].max()

In [33]:
total_process_159

Timedelta('33 days 01:44:20')

Vậy tổng thời gian sử lý của hồ sơ 159 là 33 ngày 1 giờ 44 phút 20 giây

4.	Cho biết code lý do trả về cuối cùng của các hồ sơ

In [34]:
df_create

ID APP_ID        CREATED_DATE LOAN_AMOUNT  \
13  1    123 2021-03-24 15:15:55    40000000   
14  2    159 2021-02-24 09:15:55    30000000   
15  3    173 2021-03-29 13:15:55    55000000   
16  3    168 2021-02-24 11:15:55    40000000   
17  5    761 2021-03-14 13:40:55    30000000   
18  6    234 2021-03-02 15:40:55    35000000   

                                           UW_COMMENT  month  
13  @uw01@ Trả sale bổ sung chứng từ|@uw04@ UW yêu...      3  
14                          @uw04@ UW yêu cầu bổ sung      2  
15                                                NaN      3  
16                                                NaN      2  
17  @uw06@ Trả chỉnh sửa thông tin|@uw01@ Trả sale...      3  
18                                                NaN      3

In [35]:
df_04 = df_create[df_create.UW_COMMENT.notna()]

In [36]:
df_final = df_04[['APP_ID','UW_COMMENT']]

In [37]:
df_final

APP_ID                                         UW_COMMENT
13    123  @uw01@ Trả sale bổ sung chứng từ|@uw04@ UW yêu...
14    159                          @uw04@ UW yêu cầu bổ sung
17    761  @uw06@ Trả chỉnh sửa thông tin|@uw01@ Trả sale...